### ISSUE TITLE
When new same sentence and lemma or posTag more than one , the speed will slow and memory leak  

### ISSUE THREAD

**Issue by leeguoren** [stanfordnlp/CoreNLP#287]
```
    public static void main(String[] args) {
        for (int i = 0 ; i < 1000; i++) {
            Sentence sen = new Sentence("Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. When veteran filmmaker Oliver Stone was casting an actor to play former NSA contractor-turned-whistleblower Edward Snowden in a film, he said he went to only one person - Joseph Gordon-Levitt. “I don’t know why, he just looked like, and felt like, and acted like he was one of that generation, very much the same age and computer knowledgeable,” Stone told Reuters last month in Los Angeles.    Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. Gordon-Levitt, who achieved fame as a child actor in television series 3rd Rock from the Sun, said that by playing Snowden, he hoped to understand his motivations. “I was kind of trying to figure out why he did what he did, what was going on in his head,” he said. “And one of the questions everyone asks is, ‘Why didn’t he just, you know, voice his concerns through proper channels?’” The film leads up to the events of 2013, when Snowden fled the United States after exposing the government’s mass surveillance programs to journalist Glenn Greenwald and documentary filmmaker Laura Poitras.    The US government filed espionage charges against Snowden and he was granted asylum in Russia, where he has lived since, with his girlfriend Lindsay Mills. Actor Shailene Woodley plays Mills in the film. Gordon-Levitt said he related to Snowden’s disillusionment with the US government after watching US National Intelligence director James Clapper deny, before a congressional committee, that the NSA was collecting records on millions of Americans.   “If the director of National Intelligence is being asked by a senator under oath, ‘Hey, is this happening?’ and he’s telling a lie, well, then, what is some guy that works at the NSA going to accomplish by complaining through proper channels?” the actor said. Tech-savvy Gordon-Levitt, who said he donated his fee for the film to the American Civil Liberties Union, is the founder of HitRecord, an online collaborative creative hub to brings together artists from around the world. He said he tended to be optimistic about new technology but the movie made him more aware of its negative aspects. “It’s worth being optimistic about all those things, but it’s also probably worth paying attention and considering what might the downsides be of this new technology that we’re inheriting,” he said.");
            sen.lemmas();
        }
    }
```

**Comment by manning** [comment]
What's the evidence for the memory leak?  If I run your example – I increased the iteration count to 10000 so I could see it longer and ran it with `-mx500m` – then the heap graph doesn't show any clear evidence of a memory leak to me....

![image](https://cloud.githubusercontent.com/assets/2438050/20108487/04e72b92-a591-11e6-957e-a15ebaa492af.png)

**Comment by leeguoren** [comment]
**I run it in Win7, If sentence no change, when it iteration count to 1000 use 11 minute**
```
import edu.stanford.nlp.simple.Sentence;

public class Analysis {
	public static void main(String[] args) {
		for (int i = 0 ; i < 1000; i++) {
			Sentence sen = new Sentence(" Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. When veteran filmmaker Oliver Stone was casting an actor to play former NSA contractor-turned-whistleblower Edward Snowden in a film, he said he went to only one person - Joseph Gordon-Levitt. “I don’t know why, he just looked like, and felt like, and acted like he was one of that generation, very much the same age and computer knowledgeable,” Stone told Reuters last month in Los Angeles.    Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. Gordon-Levitt, who achieved fame as a child actor in television series 3rd Rock from the Sun, said that by playing Snowden, he hoped to understand his motivations. “I was kind of trying to figure out why he did what he did, what was going on in his head,” he said. “And one of the questions everyone asks is, ‘Why didn’t he just, you know, voice his concerns through proper channels?’” The film leads up to the events of 2013, when Snowden fled the United States after exposing the government’s mass surveillance programs to journalist Glenn Greenwald and documentary filmmaker Laura Poitras.    The US government filed espionage charges against Snowden and he was granted asylum in Russia, where he has lived since, with his girlfriend Lindsay Mills. Actor Shailene Woodley plays Mills in the film. Gordon-Levitt said he related to Snowden’s disillusionment with the US government after watching US National Intelligence director James Clapper deny, before a congressional committee, that the NSA was collecting records on millions of Americans.   “If the director of National Intelligence is being asked by a senator under oath, ‘Hey, is this happening?’ and he’s telling a lie, well, then, what is some guy that works at the NSA going to accomplish by complaining through proper channels?” the actor said. Tech-savvy Gordon-Levitt, who said he donated his fee for the film to the American Civil Liberties Union, is the founder of HitRecord, an online collaborative creative hub to brings together artists from around the world. He said he tended to be optimistic about new technology but the movie made him more aware of its negative aspects. “It’s worth being optimistic about all those things, but it’s also probably worth paying attention and considering what might the downsides be of this new technology that we’re inheriting,” he said.");
			sen.lemmas();
		}
	}
}
```

![20161109103029](https://cloud.githubusercontent.com/assets/15827728/20125295/9c86f0fa-a667-11e6-8217-e7a66dfa8969.png)


**If I change the sentence everytime, it iteration count to 10000 use 1 minute**

```
import edu.stanford.nlp.simple.Sentence;

public class Analysis {
	public static void main(String[] args) {
		for (int i = 0 ; i < 10000; i++) {
			Sentence sen = new Sentence(i + " Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. When veteran filmmaker Oliver Stone was casting an actor to play former NSA contractor-turned-whistleblower Edward Snowden in a film, he said he went to only one person - Joseph Gordon-Levitt. “I don’t know why, he just looked like, and felt like, and acted like he was one of that generation, very much the same age and computer knowledgeable,” Stone told Reuters last month in Los Angeles.    Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. Gordon-Levitt, who achieved fame as a child actor in television series 3rd Rock from the Sun, said that by playing Snowden, he hoped to understand his motivations. “I was kind of trying to figure out why he did what he did, what was going on in his head,” he said. “And one of the questions everyone asks is, ‘Why didn’t he just, you know, voice his concerns through proper channels?’” The film leads up to the events of 2013, when Snowden fled the United States after exposing the government’s mass surveillance programs to journalist Glenn Greenwald and documentary filmmaker Laura Poitras.    The US government filed espionage charges against Snowden and he was granted asylum in Russia, where he has lived since, with his girlfriend Lindsay Mills. Actor Shailene Woodley plays Mills in the film. Gordon-Levitt said he related to Snowden’s disillusionment with the US government after watching US National Intelligence director James Clapper deny, before a congressional committee, that the NSA was collecting records on millions of Americans.   “If the director of National Intelligence is being asked by a senator under oath, ‘Hey, is this happening?’ and he’s telling a lie, well, then, what is some guy that works at the NSA going to accomplish by complaining through proper channels?” the actor said. Tech-savvy Gordon-Levitt, who said he donated his fee for the film to the American Civil Liberties Union, is the founder of HitRecord, an online collaborative creative hub to brings together artists from around the world. He said he tended to be optimistic about new technology but the movie made him more aware of its negative aspects. “It’s worth being optimistic about all those things, but it’s also probably worth paying attention and considering what might the downsides be of this new technology that we’re inheriting,” he said.");
			sen.lemmas();
		}
	}
}

```
![20161109103043](https://cloud.githubusercontent.com/assets/15827728/20125328/d2ef3ab2-a667-11e6-98c9-014a817d159c.png)

**How to solve the problem, when I run it in 
System: win7 
JVM: Java HotSpot(TM) 64-Bit Server VM (25.101-b13, mixed mode)
Java:  1.8.0_101**

**Comment by gangeli** [comment]
Do you get the same memory leak if you run the code through the usual pipeline (i.e., not the simple API)? And, is this using the HEAD of the GitHub repo?

**Comment by leeguoren** [comment]
it only memory leak when use simple API,  And  I use maven to import the jar.
```
    	        <dependency>
		    <groupId>edu.stanford.nlp</groupId>
		    <artifactId>stanford-corenlp</artifactId>
    		<version>3.6.0</version>
		</dependency>
		<dependency>
			<groupId>edu.stanford.nlp</groupId>
    		<artifactId>stanford-corenlp</artifactId>
    		<version>3.6.0</version>
    		<classifier>models</classifier>
		</dependency>
```

**Comment by gangeli** [comment]
I tried to replicate this bug, and got results similar to Chris. My used heap didn't appear to grow, and a dump of the heap 8 minutes apart didn't show any significant difference in the allocated classes. 

My code is:

    for (int i = 0 ; i < 100000; i++) {
      Sentence sen = new Sentence("Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. When veteran filmmaker Oliver Stone was casting an actor to play former NSA contractor-turned-whistleblower Edward Snowden in a film, he said he went to only one person - Joseph Gordon-Levitt. “I don’t know why, he just looked like, and felt like, and acted like he was one of that generation, very much the same age and computer knowledgeable,” Stone told Reuters last month in Los Angeles.    Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. Gordon-Levitt, who achieved fame as a child actor in television series 3rd Rock from the Sun, said that by playing Snowden, he hoped to understand his motivations. “I was kind of trying to figure out why he did what he did, what was going on in his head,” he said. “And one of the questions everyone asks is, ‘Why didn’t he just, you know, voice his concerns through proper channels?’” The film leads up to the events of 2013, when Snowden fled the United States after exposing the government’s mass surveillance programs to journalist Glenn Greenwald and documentary filmmaker Laura Poitras.    The US government filed espionage charges against Snowden and he was granted asylum in Russia, where he has lived since, with his girlfriend Lindsay Mills. Actor Shailene Woodley plays Mills in the film. Gordon-Levitt said he related to Snowden’s disillusionment with the US government after watching US National Intelligence director James Clapper deny, before a congressional committee, that the NSA was collecting records on millions of Americans.   “If the director of National Intelligence is being asked by a senator under oath, ‘Hey, is this happening?’ and he’s telling a lie, well, then, what is some guy that works at the NSA going to accomplish by complaining through proper channels?” the actor said. Tech-savvy Gordon-Levitt, who said he donated his fee for the film to the American Civil Liberties Union, is the founder of HitRecord, an online collaborative creative hub to brings together artists from around the world. He said he tended to be optimistic about new technology but the movie made him more aware of its negative aspects. “It’s worth being optimistic about all those things, but it’s also probably worth paying attention and considering what might the downsides be of this new technology that we’re inheriting,” he said.");
      sen.lemmas();
    }

My heap over time is:

<img width="736" alt="screen shot 2016-12-22 at 1 32 30 am" src="https://cloud.githubusercontent.com/assets/581997/21421320/1c8c20d6-c7e7-11e6-9387-3ca843ba1a9e.png">

The two heap dumps are:

<img width="1498" alt="screen shot 2016-12-22 at 1 32 48 am" src="https://cloud.githubusercontent.com/assets/581997/21421336/3099a134-c7e7-11e6-8e31-caf486d227e3.png">

<img width="1499" alt="screen shot 2016-12-22 at 1 33 07 am" src="https://cloud.githubusercontent.com/assets/581997/21421347/372e4e5a-c7e7-11e6-81e4-7d27bb3acac6.png">

Can you dump your heap at the beginning and end of your profiling period, and attach the most used classes? There's also a nifty "diff" feature you can use to compare two dumps. Randomly selecting some instances to inspect could be useful as well. From what I saw, most of the memory is in HashMap entries, and most of those entries are in the Dictionary class. If you can find a big user of memory that's not the Dictionary class, this is a good candidate for the source of the leak.

Another option: perhaps we accidentally fixed this between 3.6.0 and 3.7.0?

**Comment by leeguoren** [comment]
**I run it in Win7, If sentence no change**

```
import edu.stanford.nlp.simple.Sentence;

public class Analysis {
    public static void main(String[] args) {
        for (int i = 0 ; i < 1000; i++) {
            Sentence sen = new Sentence(" Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. When veteran filmmaker Oliver Stone was casting an actor to play former NSA contractor-turned-whistleblower Edward Snowden in a film, he said he went to only one person - Joseph Gordon-Levitt. “I don’t know why, he just looked like, and felt like, and acted like he was one of that generation, very much the same age and computer knowledgeable,” Stone told Reuters last month in Los Angeles.    Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. Gordon-Levitt, who achieved fame as a child actor in television series 3rd Rock from the Sun, said that by playing Snowden, he hoped to understand his motivations. “I was kind of trying to figure out why he did what he did, what was going on in his head,” he said. “And one of the questions everyone asks is, ‘Why didn’t he just, you know, voice his concerns through proper channels?’” The film leads up to the events of 2013, when Snowden fled the United States after exposing the government’s mass surveillance programs to journalist Glenn Greenwald and documentary filmmaker Laura Poitras.    The US government filed espionage charges against Snowden and he was granted asylum in Russia, where he has lived since, with his girlfriend Lindsay Mills. Actor Shailene Woodley plays Mills in the film. Gordon-Levitt said he related to Snowden’s disillusionment with the US government after watching US National Intelligence director James Clapper deny, before a congressional committee, that the NSA was collecting records on millions of Americans.   “If the director of National Intelligence is being asked by a senator under oath, ‘Hey, is this happening?’ and he’s telling a lie, well, then, what is some guy that works at the NSA going to accomplish by complaining through proper channels?” the actor said. Tech-savvy Gordon-Levitt, who said he donated his fee for the film to the American Civil Liberties Union, is the founder of HitRecord, an online collaborative creative hub to brings together artists from around the world. He said he tended to be optimistic about new technology but the movie made him more aware of its negative aspects. “It’s worth being optimistic about all those things, but it’s also probably worth paying attention and considering what might the downsides be of this new technology that we’re inheriting,” he said.");
            sen.lemmas();
        }
    }
}
```

![testtwogc](https://cloud.githubusercontent.com/assets/15827728/21446525/28f5f8bc-c902-11e6-9d51-42bb379f0028.png)
![testtwo](https://cloud.githubusercontent.com/assets/15827728/21446555/8a111352-c902-11e6-91f7-c06452d221ed.png)
![testtwoclass](https://cloud.githubusercontent.com/assets/15827728/21446558/9159f5fc-c902-11e6-850a-c75bbaf5c93e.png)
![testtwodump](https://cloud.githubusercontent.com/assets/15827728/21446559/956469c0-c902-11e6-900f-fdd0d8e6cc0e.png)









**If I change the sentence everytime**

```
import edu.stanford.nlp.simple.Sentence;

public class Analysis {
    public static void main(String[] args) {
        for (int i = 0 ; i < 10000; i++) {
            Sentence sen = new Sentence(i + " Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. When veteran filmmaker Oliver Stone was casting an actor to play former NSA contractor-turned-whistleblower Edward Snowden in a film, he said he went to only one person - Joseph Gordon-Levitt. “I don’t know why, he just looked like, and felt like, and acted like he was one of that generation, very much the same age and computer knowledgeable,” Stone told Reuters last month in Los Angeles.    Snowden, which had its world premiere on Friday at the Toronto International Film Festival and hits theatres on September 16, sees Gordon-Levitt, 35, play the 33-year-old Snowden through a decade of his life. Gordon-Levitt, who achieved fame as a child actor in television series 3rd Rock from the Sun, said that by playing Snowden, he hoped to understand his motivations. “I was kind of trying to figure out why he did what he did, what was going on in his head,” he said. “And one of the questions everyone asks is, ‘Why didn’t he just, you know, voice his concerns through proper channels?’” The film leads up to the events of 2013, when Snowden fled the United States after exposing the government’s mass surveillance programs to journalist Glenn Greenwald and documentary filmmaker Laura Poitras.    The US government filed espionage charges against Snowden and he was granted asylum in Russia, where he has lived since, with his girlfriend Lindsay Mills. Actor Shailene Woodley plays Mills in the film. Gordon-Levitt said he related to Snowden’s disillusionment with the US government after watching US National Intelligence director James Clapper deny, before a congressional committee, that the NSA was collecting records on millions of Americans.   “If the director of National Intelligence is being asked by a senator under oath, ‘Hey, is this happening?’ and he’s telling a lie, well, then, what is some guy that works at the NSA going to accomplish by complaining through proper channels?” the actor said. Tech-savvy Gordon-Levitt, who said he donated his fee for the film to the American Civil Liberties Union, is the founder of HitRecord, an online collaborative creative hub to brings together artists from around the world. He said he tended to be optimistic about new technology but the movie made him more aware of its negative aspects. “It’s worth being optimistic about all those things, but it’s also probably worth paying attention and considering what might the downsides be of this new technology that we’re inheriting,” he said.");
            sen.lemmas();
        }
    }
}

```
![testonegc](https://cloud.githubusercontent.com/assets/15827728/21446552/7ba222ca-c902-11e6-89a5-9177444afcaf.png)
![testone](https://cloud.githubusercontent.com/assets/15827728/21446528/34ca14de-c902-11e6-8c20-14ae031adc47.png)
![testoneclass](https://cloud.githubusercontent.com/assets/15827728/21446533/41405156-c902-11e6-998f-28e14b30e39f.png)
![testonedump](https://cloud.githubusercontent.com/assets/15827728/21446539/532fbc12-c902-11e6-849a-17f6837be4e5.png)

**How to solve the problem, when I run it in 
System: win7 
JVM: Java HotSpot(TM) 64-Bit Server VM (25.101-b13, mixed mode)
Java:  1.8.0_101**

If I use  3.7.0,  heap didn't appear to grow

**Comment by gangeli** [comment]
Strange. Perhaps this was related to an annotator caching bug fixed between 3.6.0 and 3.7.0? Or I was caching annotations incorrectly in the Document class itself? But, if it works correctly in 3.7.0, is it safe if I close this as "mysteriously solved itself?"

**Comment by mattilyra** [comment]
I've experienced a similar issue with the CoreNLPServer. I've been running it for about two weeks now in a docker container and the container memory consumption has gone from about 1.5Gb to 2.9Gb. The CoreNLPServer is the only thing running in the container. I'm using the latest 3.7.0 binaries from (`2016-10-31`).

**Comment by gangeli** [comment]
What's the max memory you're giving the Java process? The most likely cause of memory leaks is the caching of annotators (each different set of properties passed into the server effectively creates different annotators -- to prevent a huge overhead every time you make a call, we cache these). But, the cache _should_ be bounded by the memory given to the Java process, and evict old annotators as needed.

A differential Heap dump between the process ~1 day in and the process a week or two in would be most useful to help debug the issue.

**Comment by mattilyra** [comment]
I'm starting to think this is an issue with `docker` not with CoreNLP. I figured out a way to bring down the process faster and started a separate process without `docker` for testing that, that process however seems to be running stable. I think the issue might be related to docker and logs, still looking into that - I've redirected the `stderr` and `stdout` to a file now so we'll see if that fixes it (for `docker`).  

I have diff heap dumps from the non-docker Java process if you're interested but it might be a waste of your time. The command I'm using the run start corenlp is

> java -mx4g -cp "*" edu.stanford.nlp.pipeline.StanfordCoreNLPServer -threads 4 -annotators tokenize,ssplit,pos,ner -ner.model edu/stanford/nlp/models/ner/english.all.3class.distsim.crf.ser.gz -ner.useSUTime false -ner.applyNumericClassifiers false > /dev/null 2> ./stderr.log

At the moment, after about 24hrs of servic HTTP requests, the process has taken up all of that 4G but it's still running fine.
